### ISSUE TITLE
Chapters 1-3: Assignement

### ISSUE THREAD

**Issue by radumarias** [xoriors/course-rust#5]
Please see [The Rust Book](https://doc.rust-lang.org/book/) and follow the steps from [For students](https://github.com/xoriors/course-rust/blob/main/docs/students.md) to create and submit your solutions.

Chapters 1-3 #4

Please create the code needed for this and submit it as above.

# Crypto mining

This is how crypto mining with proof of work works, in principle.

Read a number from the keyboard; let's call it `N.`  between `1` and `100` (you can use bigger numbers if you want, but it will take considerably longer).  
Generate `N` random numbers between `0.0 and 1.0` as `f32` (you can play with less precision ones in order to reduce the time) and keep them in a `Vec` or an array `[f32; N]`.  
Now, in a loop, keep generating random numbers and remove them from the list or mark the ones that match. Count how many times you have to generate until you remove all from the list as, aka you generate all of those again.  
Repeat the experiment let's say 100 times for each N, and print the `min,` `avg,` and `max` of times you had to generate.

# Advanced, not necessarily, only if you are very adventurous :)
# Crypto mining with [weighted random distribution](https://chatgpt.com/share/67b20ed6-0114-8003-ae19-eb4bfe9a42af)

Now repeat the above, but instead of random numbers between `0 and 1` as float, generate random int greater than `0` in the `0..1 billion` interval.  
Then calculate the weight of each interval of 100 numbers, such as [1..100], [101,199], ... The weight is calculated by how many numbers from those random N ones are in the interval.  
Now, in the loop where you try to generate all the numbers in the array, use weighted random distribution (to generate some of the  random numbers with a higher probability).  
You can see more on [weighted random distribution](https://docs.rs/rand_distr/latest/rand_distr/weighted/index.html). You can use functions from there and then try to implement your own .

tl;dr

weight can be easily applied to specific numbers, not interval
so I tried to split the range in intervals and consider the offset of the interval as the value to which we apply the weight, like that somehow we give weight to the intervals

to calculate the weight for an interval we calculate how many of the numbers were generated in that interval, so the weight is that count

then use a weighted random generator to generate the index/offset of the interval (this will ensure intervals that contains more numbers will be generated more often, based on the weight)
then generate a random number in his interval marking the hits in the initial random number list

this somehow ensures we generate more often numbers in more populated intervals 

hint: as we mark/remove numbers we hit, the remaining numbers in the interval decreases and normally another interval could have more numbers that this one, so normally the weights changes 
if we want to maximize the numbers we need to generate to hit all remaining numbers we would recalculate the weights after each hit

# We also recommend [Learn by doing](https://github.com/xoriors/course-rust/wiki/Additional-resources#learn-by-doing)
